### ISSUE TITLE
Consider introducing utility method for checked functions

### ISSUE THREAD

**Issue by dfeist** [reactor/reactor-core#80]
When adopting reactor, but needing to wrap and use existing code you find yourself using try/catch blocks within `map` and other operators repeatedly which severely impacts the readability of code.  The introduction of a CheckedFunction could help with this significantly.

`Exceptions.checked(CheckedFunction) : Function`

**Comment by smaldini** [comment]
I'm not too keen on expanding that API surface easily but wdyt @akarnokd @rstoyanchev @sdeleuze ?

**Comment by dfeist** [comment]
I agree on not adding this to Flux/Mono, but maybe something outside Flux.  With this one it's a good idea to think about how wrapping of existing blocking API's will be documented, what approach will be suggested and what example would be provided..

**Comment by sdeleuze** [comment]
What about just using `Exceptions.propagate(Throwable t)` and optionally the newly introduced `Flux.mapError(Function<Throwable, Throwable> mapper)`?

**Comment by dfeist** [comment]
That works yes, this isn't something you can't do, rather an usability/readability improvement suggestion.  try/catch look very ugly and affect readability, the idea here is to simply hide the `try/catch/Exceptions.propagate(t)` in something reusable.

**Comment by smaldini** [comment]
To evaluate in an extra addon for blocking utils like reactor-blocking

**Comment by dfeist** [comment]
I've been thinking about this one and playing around with some options. Even though i agreed with @smaldini yesterday regarding leaving this out of 3.0 and implementing in a different module.  After playing around with implementation and considering:
i) It doesn't affect Flux API surface area
ii) the number of people that will use reactor with existing code that throws checked exceptions will be significant
iii) the improvement to usability/readability when using this approach if significant
I think it's a no-brainer to add this to Exceptions.  (can use a different name instead of wrap if it makes sense).

**Before:**
```
        Flux.range(0, 100).map(v -> {
                                   try {
                                       return transform(v);
                                   } catch (Exception e) {
                                       throw propagate(e);
                                   }
                               })
                          .subscribe(v -> {
                               try {
                                   log(v);
                               } catch (Exception e) {
                                   throw propagate(e);
                               }
                           });
```
**After:**
```
        Flux.range(0, 100).map(wrap(v -> transform(v))).subscribe(wrap(v -> log(v)));
```

**Comment by smaldini** [comment]
I'd be explicit on the wrap method names like `checkedConsumer(CheckedConsumer)` etc, I'm not sure wrap will not conflict in static import anyway. Other than that WDYT @akarnokd @rstoyanchev @sdeleuze ? It would add a few helper functional classes under Exceptions to the surface API (CheckedConsumer, CheckedBiConsumer, CheckedBiFunction, CheckedFunction, CheckedPredicate, CheckedBiPredicate). I can still see a new module for this.

**Comment by dfeist** [comment]
@smaldini I'm not convinced a separate module is needed as core is already dealing with exception wrapping/propagation/unwrapping and this is but I'm not against a separate module if others think that is best. I do feel that this should be available for 3.0.0 though, for both usability and adoption as a lot of users will need to integrate with existing methods throwing checked exceptions. And just look at the readability and verbosity of the before/after above..

Regarding naming, I wasn't convinced about `wrap` either, but i do feel quite strongly that overloaded methods should be used for usability and reduced verbosity rather than different method names per type.  I did add the the helper functional classes under Exceptions you mentioned, just not all of them.

I just pushed updated naming and additional functions for review..
